airflow/models/mappedoperator.py

```diff
@@ -686,33 +693,40 @@ def render_template_fields(
         """
         if not jinja_env:
             jinja_env = self.get_template_env()
-        unmapped_task = self.unmap()
+        # Before we unmap we have to resolve the mapped arguments, otherwise the real operator constructor
+        # could be called with an XComArg, rather than the value it resolves to.
+        #
+        # We also need to resolve _all_ mapped arguments, even if they aren't marked as templated
+        kwargs = self._get_unmap_kwargs()
+
+        template_fields = set(self.template_fields)
+
+        self._resolve_expansion_kwargs(kwargs, template_fields, context, session)
+
+        unmapped_task = self.unmap(rendered_kwargs=kwargs)
         self._do_render_template_fields(
             parent=unmapped_task,
-            template_fields=unmapped_task.template_fields,
+            template_fields=template_fields,
             context=context,
             jinja_env=jinja_env,
             seen_oids=set(),
         )
         return unmapped_task
 
-    def _render_template_field(
```

Comment:
I think we can remove this in AbstractOperator now; the sole reason of this function’s existence if for MappedOperator to hook into.

Reply:
Oh missed that. Cool.